README.md
# guh-plugin-buderus

Plugin for the [guh](https://github.com/guh/guh) IoT platform that
allows to connect to Buderus heating systems (via Buderus KM200 Web Gateway).

## Command Line Client

The project contains a command line client `scripts/buderus-km200-cli.py`
that can either query an actual gateway or act as a web server and emulate a
gateway. This is useful to test the plugin without connecting to real hardware.

Query mode:

    ./buderus-km200-cli.py km200 <KEY> gw.json

Emulation mode:

    ./buderus-km200-cli.py --mode emulate localhost:5000 <KEY> gw.json

## License

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
[GNU General Public License](http://www.gnu.org/licenses/gpl-2.0.html>)
for more details.


debian/changelog
guh-plugin-buderus (0.1) UNRELEASED; urgency=medium

  [ Christian Fetzer ]
  * Initial release. (Closes: #XXXXXX)

 --  <fetzerch@gmail.com>  Sun, 05 Nov 2017 17:12:00 +0000


devicepluginbuderus.h
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *                                                                         *
 *  Copyright (C) 2017 Christian Fetzer <fetzer.ch@gmail.com>              *
 *                                                                         *
 *  This file is part of guh.                                              *
 *                                                                         *
 *  Guh is free software: you can redistribute it and/or modify            *
 *  it under the terms of the GNU General Public License as published by   *
 *  the Free Software Foundation, version 2 of the License.                *
 *                                                                         *
 *  Guh is distributed in the hope that it will be useful,                 *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of         *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the           *
 *  GNU General Public License for more details.                           *
 *                                                                         *
 *  You should have received a copy of the GNU General Public License      *
 *  along with guh. If not, see <http://www.gnu.org/licenses/>.            *
 *                                                                         *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

#ifndef DEVICEPLUGINBUDERUS_H
#define DEVICEPLUGINBUDERUS_H

#include "plugin/deviceplugin.h"
#include "devicemanager.h"

class DevicePluginBuderus: public DevicePlugin
{
    Q_OBJECT

    Q_PLUGIN_METADATA(IID "guru.guh.DevicePlugin" FILE "devicepluginbuderus.json")
    Q_INTERFACES(DevicePlugin)

public:
    DeviceManager::HardwareResources requiredHardware() const override;
    DeviceManager::DeviceSetupStatus setupDevice(Device *device) override;

    void guhTimer() override;
    void networkManagerReplyReady(QNetworkReply *reply) override;

    void deviceRemoved(Device *device) override;

private:
    struct Request {
        enum class Type {
            Unknown,
            Param,
            State
        };

        Request(const char *url, const ParamTypeId &paramTypeId, Device *device)
            : url{url}, param{paramTypeId}, device{device}, type{Type::Param}
        {}

        Request(const char *url, const StateTypeId &stateTypeId, Device *device)
            : url{url}, state{stateTypeId}, device{device}, type{Type::State}
        {}

        Request() = default;

        QString url;
        ParamTypeId param;
        StateTypeId state;
        Device *device = nullptr;
        Type type = Type::Unknown;
    };

    void sendAsyncRequest(const Request &request);
    QVariant parseValue(Device *device, const QByteArray &responseText,
                        const QString &key = QStringLiteral("value"));
    static QByteArray decrypt(const QByteArray &encrypted, const QByteArray &key);

private:
    QMap<QNetworkReply*, Request> m_asyncRequests;
};

#endif // DEVICEPLUGINBUDERUS_H


devicepluginbuderus.cpp
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *                                                                         *
 *  Copyright (C) 2017 Christian Fetzer <fetzer.ch@gmail.com>              *
 *                                                                         *
 *  This file is part of guh.                                              *
 *                                                                         *
 *  Guh is free software: you can redistribute it and/or modify            *
 *  it under the terms of the GNU General Public License as published by   *
 *  the Free Software Foundation, version 2 of the License.                *
 *                                                                         *
 *  Guh is distributed in the hope that it will be useful,                 *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of         *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the           *
 *  GNU General Public License for more details.                           *
 *                                                                         *
 *  You should have received a copy of the GNU General Public License      *
 *  along with guh. If not, see <http://www.gnu.org/licenses/>.            *
 *                                                                         *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

#include "devicepluginbuderus.h"
#include "plugininfo.h"

#include "Qt-AES/qaesencryption.h"

#include <QJsonDocument>

namespace {
const char *uuidUrl = "/gateway/uuid";
const char *versionFirmwareUrl = "/gateway/versionFirmware";
const char *temperatureOutdoorUrl = "/system/sensors/temperatures/outdoor_t1";
}

DeviceManager::HardwareResources DevicePluginBuderus::requiredHardware() const
{
    return DeviceManager::HardwareResourceNetworkManager | DeviceManager::HardwareResourceTimer;
}

DeviceManager::DeviceSetupStatus DevicePluginBuderus::setupDevice(Device *device)
{
    if (device->deviceClassId() != buderusGatewayDeviceClassId) {
        return DeviceManager::DeviceSetupStatusFailure;
    }

    auto host = device->paramValue(buderusHostParamTypeId).toString();
    auto port = device->paramValue(buderusPortParamTypeId).toString();
    qCDebug(dcBuderus) << "Configuring device" << QString("http://%1:%2").arg(host).arg(port);

    sendAsyncRequest(Request{uuidUrl, buderusIdParamTypeId, device});
    sendAsyncRequest(Request{versionFirmwareUrl, buderusVersionParamTypeId, device});
    return DeviceManager::DeviceSetupStatusAsync;
}

void DevicePluginBuderus::guhTimer()
{
    foreach (Device *device, myDevices()) {
        if (!device->setupComplete())
            continue;

        sendAsyncRequest(Request{temperatureOutdoorUrl, temperatureOutdoorStateTypeId, device});
    }
}

void DevicePluginBuderus::sendAsyncRequest(const Request &request)
{
    auto host = request.device->paramValue(buderusHostParamTypeId).toString();
    auto port = request.device->paramValue(buderusPortParamTypeId).toString();
    QNetworkRequest networkRequest(QUrl(QString("http://%1:%2%3").arg(host).arg(port).arg(request.url)));
    networkRequest.setHeader(QNetworkRequest::UserAgentHeader, "TeleHeater/2.2.3");
    networkRequest.setRawHeader("Accept", "application/json");
    qCDebug(dcBuderus) << "Sending request:" << request.url;
    auto reply = networkManagerGet(networkRequest);
    m_asyncRequests.insert(reply, request);
}

void DevicePluginBuderus::networkManagerReplyReady(QNetworkReply *reply)
{
    if (!m_asyncRequests.contains(reply))
        return;

    auto request = m_asyncRequests.take(reply);
    if (!reply->error()) {
        auto value = parseValue(request.device, reply->readAll());
        qCDebug(dcBuderus) << "Received value:" << request.url << value;
        if (request.type == Request::Type::Param) {
            request.device->setParamValue(request.param, value);
            if (request.url == versionFirmwareUrl) {
                emit deviceSetupFinished(request.device, DeviceManager::DeviceSetupStatusSuccess);
            }
        } else if (request.type == Request::Type::State) {
            request.device->setStateValue(request.state, value);
        }

        // Update the timestamp only for 1 value per request type.
        if (request.url == temperatureOutdoorUrl || request.url == versionFirmwareUrl) {
            request.device->setStateValue(updateTimeStateTypeId, QDateTime::currentMSecsSinceEpoch() / 1000);
        }
    } else {
        qCWarning(dcBuderus) << "Reply error" << reply->errorString();
        if (request.url == versionFirmwareUrl) {
            emit deviceSetupFinished(request.device, DeviceManager::DeviceSetupStatusFailure);
        }
    }
    reply->deleteLater();
}

void DevicePluginBuderus::deviceRemoved(Device *device)
{
    // Remove pending requests.
    for (auto it = m_asyncRequests.begin(); it != m_asyncRequests.end();) {
        if (it.value().device->id() == device->id()) {
            it.key()->deleteLater();
            it = m_asyncRequests.erase(it);
        } else {
            ++it;
        }
    }
}

QVariant DevicePluginBuderus::parseValue(Device *device, const QByteArray &responseText,
                                         const QString &key)
{
    QByteArray decryptionKey = device->paramValue(buderusKeyParamTypeId).toString().toLocal8Bit();
    auto text = decrypt(responseText, decryptionKey);
    auto json = QJsonDocument::fromJson(text);
    auto result = json.object().value(key).toVariant();
    return result;
}

QByteArray DevicePluginBuderus::decrypt(const QByteArray &encrypted, const QByteArray &key)
{
    QAESEncryption encryption{QAESEncryption::AES_256, QAESEncryption::ECB};
    auto base64decoded = QByteArray::fromBase64(encrypted);
    auto decrypted = encryption.decode(base64decoded, QByteArray::fromHex(key));
    while (decrypted.endsWith('\0')) decrypted.chop(1);
    return decrypted;
}


buderus-wireshark/init.lua
-- Import dependencies.
local gcrypt
do
    local status, res = pcall(require, "luagcrypt")
    if status then
        gcrypt = res
    else
        report_failure("Failed to load Luagcrypt. See README.md.\n" .. res)
    end
end
local json = require "json"

local function buderus_dissect(proto, tvbuffer, treeitem)
    local subtree = treeitem:add(proto, tvbuffer(), "Buderus KM200 Packet")
    local payload = tvbuffer():string()

    -- Remove preceding HTTP header and following XMPP/XML tags.
    local hdr_len = payload:find("\r\n\r\n")
    if hdr_len == nil then
        return
    end
    local content = payload:sub(hdr_len + 4)
    local index = content:find('<')
    if index ~= nil then
        content = content:sub(0, content:find('<') - 1)
    end
    subtree:add("Data: " .. content)

    -- Decrypt content.
    local cipher = gcrypt.Cipher(gcrypt.CIPHER_AES256, gcrypt.CIPHER_MODE_ECB)
    if not pcall(function()
        cipher:setkey(Struct.fromhex(proto.prefs.key))
    end) then
        subtree:add("Invalid decryption key set in protocol preferences.")
        return
    end
    local decoded = ByteArray.new(content, true):base64_decode()
    local decrypted = cipher:decrypt(decoded:raw())
    subtree:add("Decrypted Data: " .. decrypted)

    -- Parse the decrypted packet as JSON.
    local json_obj
    if not pcall(function()
        json_obj = json.decode(string.match(decrypted, "{.*}"))
    end) then
        return
    end
    local subtree2 = subtree:add("JSON Contents:")
    local function json_add(object, tree)
        for key, value in pairs(object) do
            if not(type(value) == "table") then
                tree:add(key .. ": " .. value)
            else
                subtree = tree:add(key .. ":")
                json_add(value, subtree)
            end
        end
    end
    json_add(json_obj, subtree2)
end

local buderus_http_proto = Proto("buderus_http","Buderus KM200 (HTTP)")
buderus_http_proto.prefs.key = Pref.string( "Key", "", "AES256 Decryption Key")
local original_http_dissector
function buderus_http_proto.dissector(tvbuffer, pinfo, treeitem)
    -- Call stored http dissector so that its output is available as well.
    -- https://osqa-ask.wireshark.org/questions/57685/reassembly-not-working-with-a-chained-dissector-in-lua
    local can_desegment_saved = pinfo.can_desegment
    if pinfo.can_desegment > 0 then
        pinfo.can_desegment = 2
    end
    original_http_dissector:call(tvbuffer, pinfo, treeitem)
    pinfo.can_desegment = can_desegment_saved

    buderus_dissect(buderus_http_proto, tvbuffer, treeitem)
end

local buderus_xmpp_proto = Proto("buderus_xmpp","Buderus KM200 (XMPP)")
buderus_xmpp_proto.prefs.key = Pref.string( "Key", "", "AES256 Decryption Key")
local original_xmpp_dissector
function buderus_xmpp_proto.dissector(tvbuffer, pinfo, treeitem)
    -- Call stored xmpp dissector so that its output is available as well.
    -- https://osqa-ask.wireshark.org/questions/57685/reassembly-not-working-with-a-chained-dissector-in-lua
    local can_desegment_saved = pinfo.can_desegment
    if pinfo.can_desegment > 0 then
        pinfo.can_desegment = 2
    end
    original_xmpp_dissector:call(tvbuffer, pinfo, treeitem)
    pinfo.can_desegment = can_desegment_saved

    buderus_dissect(buderus_xmpp_proto, tvbuffer, treeitem)
end

local tcp_dissector_table = DissectorTable.get("tcp.port")
original_http_dissector = Dissector.get("http")
original_xmpp_dissector = Dissector.get("xmpp")
tcp_dissector_table:add(80, buderus_http_proto)
tcp_dissector_table:add(5222, buderus_xmpp_proto)


scripts/buderus-km200-cli.py
#!/usr/bin/env python3

import argparse
import base64
import codecs
import json
import sys
import time
from http.server import BaseHTTPRequestHandler, HTTPServer

import requests
from Crypto.Cipher import AES


def decrypt(ciphertext, key):
    cipher = AES.new(codecs.decode(key, 'hex_codec'), AES.MODE_ECB)
    result = cipher.decrypt(base64.b64decode(ciphertext))
    result = result.rstrip(b'\x00').decode('utf-8')
    return result


def encrypt(plaintext, key):
    plaintext = plaintext.encode('utf-8')
    plaintext = plaintext + \
                (AES.block_size - len(plaintext) % AES.block_size) * b'\x00'
    cipher = AES.new(codecs.decode(key, 'hex_codec'), AES.MODE_ECB)
    result = cipher.encrypt(plaintext)
    result = base64.b64encode(result)
    return result.decode('utf-8')


def test_encryption():
    plaintext = '{"id":"/gateway/DateTime","type":"stringValue",' \
                '"writeable":1,"recordable":0,"value":"2017-07-21T17:15:00"}'
    key = '11111111111111111111111111111111' \
          '11111111111111111111111111111111'
    ciphertext = 'lSNasjLWs9oUC0hkNB0AcMDV+jQobTQD7KwHjDDfv8GeDKNR+TiFG8wFf' \
                 'DAiCjlnSUqLxNa+RWPF9upIJzQ01M6vtBpJWUUvGfG8WH8fJRt0WKiMbj' \
                 '914aS7pEdWQGuC1+WGKVvCFAN+Nz2bAezNpA=='
    encrypted = encrypt(plaintext, key)
    assert encrypted == ciphertext
    decrypted = decrypt(ciphertext, key)
    assert decrypted == plaintext


class BuderusServer(BaseHTTPRequestHandler):
    CONTENT = {}
    KEY = ''

    def do_GET(self):
        if self.path in self.CONTENT:
            self.send_response(200)
            self.send_header("Content-type", "application/json")
            self.end_headers()
            data = json.dumps(self.CONTENT[self.path])
            self.wfile.write(encrypt(data, self.KEY).encode('utf-8'))
        else:
            self.send_error(404, 'file not found')


def get_value(url, settings):
    headers = {'User-Agent': 'TeleHeater/2.2.3', 'Accept': 'application/json'}
    print("Requesting {}".format(url))
    try:
        response = requests.get(
            'http://{}{}'.format(settings.host, url), headers=headers)
        if response.status_code == 200:
            try:
                data = json.loads(decrypt(response.text, settings.key))
                return data
            except Exception as err:
                print("Failed to parse: {}: {}".format(url, err))
                return None
        else:
            print("Failed to get value: {}".format(url))
            return None
    except requests.RequestException as err:
        raise IOError


def get_values_recursive(url, settings, result):
    data = get_value(url, settings)
    if data:
        result[url] = data
        if 'references' in data:
            for key in data['references']:
                time.sleep(0.1)
                get_values_recursive(key['id'], settings, result)


def query(settings):
    urls = ['/system', '/gateway', '/heatSources', '/heatingCircuits',
            '/dhwCircuits', '/solarCircuits', '/recordings', '/notifications']
    if settings.args:
        urls = settings.args
    result = {}
    print("Please wait, this might take a bit...")
    try:
        for url in urls:
            get_values_recursive(url, settings, result)
    except KeyboardInterrupt:
        print("Interrupted")
    except IOError:
        print("Connection error")

    json_settings = {'sort_keys': True, 'indent': 4, 'separators': (',', ': ')}
    if settings.file:
        with open('out.json', 'w') as outfile:
            json.dump(result, outfile, **json_settings)
    else:
        print(json.dumps(result, **json_settings))


def emulate(settings):
    data = {}
    with open(settings.file) as data_file:
        data = json.load(data_file)
    server = HTTPServer(settings.host, BuderusServer)
    BuderusServer.CONTENT = data
    BuderusServer.KEY = settings.key
    print("Server listening on {}:{} ...".format(
        settings.host[0], settings.host[1]))
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        pass
    server.server_close()


def main():
    parser = argparse.ArgumentParser(description='Buderus KM200 Utility')
    parser.add_argument('--mode', choices=['query', 'emulate'],
                        default='query',
                        help="query: connect to real Buderus KM200 Gateway; "
                             "emulate: emulate a Buderus KM200 Gateway")
    parser.add_argument('host',
                        help="query: connect to host; emulate: bind to host")
    parser.add_argument('key',
                        help="en/decryption key")
    parser.add_argument('file', nargs='?',
                        help="query: write output to file (default: stdout); "
                             "emulate: read input from file")
    parser.add_argument('--args', nargs=argparse.REMAINDER)
    settings = parser.parse_args()
    print(settings)

    if len(settings.key) != 64:
        parser.error("key must be 64 hex characters (256 bit)")
    try:
        int(settings.key, 16)
    except ValueError:
        parser.error("key must be 64 hex characters (256 bit)")
    if settings.mode == 'query':
        if not settings.host:
            parser.error("query mode requires host parameter")
        query(settings)
    elif settings.mode == 'emulate':
        if ':' in settings.host:
            host = settings.host.rsplit(':')
            settings.host = (host[0], int(host[1]))
        else:
            settings.host = (settings.host, int(5000))
        if not settings.file:
            parser.error("emulate mode requires file parameter")
        emulate(settings)


if __name__ == '__main__':
    main()
